core/src/main/java/org/infinispan/commands/impl/InvocationManagerImpl.java

```diff
@@ -0,0 +1,185 @@
+         } else {
+            keys[segment][pos] = key;
+            ids[segment][pos] = commandInvocationId.getId();
+            lengths[segment] = pos + 1;
+         }
+      }
+      forgetInvocations(segment, fic);
+   }
+
+   @Override
+   public void notifyCompleted(CommandInvocationId commandInvocationId, Collection<?>[] keysBySegment) {
+      assert rpcManager.getAddress().equals(commandInvocationId.getAddress());
+      for (int segment = 0; segment < keysBySegment.length; ++segment) {
+         Collection<?> commandKeys = keysBySegment[segment];
+         if (commandKeys == null) {
+            continue;
+         }
+         ForgetInvocationsCommand fic = null;
+         synchronized (locks[segment]) {
+            int pos = lengths[segment];
+            if (trace) {
+               log.tracef("Segment %d has currently %d completed commands, another %d are being completed", segment, pos, commandKeys.size());
+            }
+            int segmentKeys = pos + commandKeys.size();
+            if (segmentKeys > MAX) {
+               fic = cf.buildForgetInvocationsCommand();
+               Object[] keys = Arrays.copyOf(this.keys[segment], segmentKeys);
+               long[] ids = Arrays.copyOf(this.ids[segment], MAX + 1);
+               for (Object key : commandKeys) {
+                  keys[pos] = key;
+                  ids[pos] = commandInvocationId.getId();
+                  ++pos;
+               }
+               fic.set(keys, ids);
+               // we fill the array to release keys for GC, but don't have to zero ids
+               Arrays.fill(this.keys[segment], 0, lengths[segment], null);
+               lengths[segment] = 0;
+            } else {
+               Object[] keys = this.keys[segment];
+               long[] ids = this.ids[segment];
+               for (Object key : commandKeys) {
+                  keys[pos] = key;
+                  ids[pos] = commandInvocationId.getId();
+                  ++pos;
+               }
+               lengths[segment] = segmentKeys;
+            }
+         }
+         forgetInvocations(segment, fic);
+      }
+   }
+
+   private void forgetInvocations(int segment, ForgetInvocationsCommand fic) {
+      if (fic != null) {
+         DistributionInfo info = dm.getCacheTopology().getDistributionForSegment(segment);
+         rpcManager.invokeRemotelyAsync(info.writeOwners(), fic, forgetRpcOptions);
+         if (info.isWriteOwner()) {
+            fic.init(dataContainer);
+            fic.setOrigin(rpcManager.getAddress());
+            if (trace) {
+               log.tracef("Invoking %s", fic);
+            }
+            fic.invoke();
+            if (trace) {
+               Set<Object> keys = new HashSet<>(Arrays.asList(fic.keys()));
+               log.tracef("Invalidation completed: " + StreamSupport.stream(dataContainer.spliterator(), false)
+                     .filter(ice -> keys.contains(ice.getKey())).collect(Collectors.toList()));
```

Comment:
Minor, but you can use [1] instead to avoid the List

[1] https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#joining--